File: src/clone_wars/web/render/format.py
from __future__ import annotations

from clone_wars.engine.types import ObjectiveStatus, SQUAD_SIZE, Supplies, UnitStock


def pct(value: float) -> int:
    return int(max(0.0, min(1.0, value)) * 100)


def fmt_int(value: int) -> str:
    return f"{value:,}"


def fmt_squads(squads: int) -> str:
    troops = squads * SQUAD_SIZE
    return f"{fmt_int(squads)} sq ({fmt_int(troops)})"


def bar(value: int, max_value: int, width: int = 18) -> str:
    if max_value <= 0:
        max_value = 1
    value = max(0, value)
    filled = int(min(1.0, value / max_value) * width)
    return "[" + ("=" * filled) + ("." * (width - filled)) + "]"


def status_label(status: ObjectiveStatus) -> str:
    if status == ObjectiveStatus.ENEMY:
        return "ENEMY HELD"
    if status == ObjectiveStatus.CONTESTED:
        return "CONTESTED"
    return "FRIENDLY"


def status_class(status: ObjectiveStatus) -> str:
    if status == ObjectiveStatus.ENEMY:
        return "status-enemy"
    if status == ObjectiveStatus.CONTESTED:
        return "status-contested"
    return "status-secured"


def risk_label(risk: float) -> str:
    if risk <= 0.0:
        return "SECURE"
    if risk <= 0.015:
        return "LOW"
    if risk <= 0.04:
        return "ELEVATED"
    return "HIGH"


def sum_supplies(stocks: dict[object, Supplies]) -> Supplies:
    ammo = sum(s.ammo for s in stocks.values())
    fuel = sum(s.fuel for s in stocks.values())
    med = sum(s.med_spares for s in stocks.values())
    return Supplies(ammo=ammo, fuel=fuel, med_spares=med)


def sum_units(units: dict[object, UnitStock]) -> UnitStock:
    infantry = sum(u.infantry for u in units.values())
    walkers = sum(u.walkers for u in units.values())
    support = sum(u.support for u in units.values())
    return UnitStock(infantry=infantry, walkers=walkers, support=support)

File: src/clone_wars/frontend/src/utils/format.ts
import type { Depot, Supplies, UnitStock } from "../api/types";

export function fmtInt(value: number): string {
  return Intl.NumberFormat("en-US").format(value);
}

export function fmtPct(value: number, digits = 0): string {
  return `${(value * 100).toFixed(digits)}%`;
}

export function sumSupplies(depots: Depot[]): Supplies {
  return depots.reduce(
    (acc, depot) => ({
      ammo: acc.ammo + depot.supplies.ammo,
      fuel: acc.fuel + depot.supplies.fuel,
      medSpares: acc.medSpares + depot.supplies.medSpares
    }),
    { ammo: 0, fuel: 0, medSpares: 0 }
  );
}

export function sumUnits(depots: Depot[]): UnitStock {
  return depots.reduce(
    (acc, depot) => ({
      infantry: acc.infantry + depot.units.infantry,
      walkers: acc.walkers + depot.units.walkers,
      support: acc.support + depot.units.support
    }),
    { infantry: 0, walkers: 0, support: 0 }
  );
}

export function labelize(id: string): string {
  return id
    .replace("contested_", "")
    .replace("new_system_", "")
    .replace(/_/g, " ")
    .toUpperCase();
}
